### PR TITLE
feature: add a mechanism for clients to store data on the server

### PR DESCRIPTION
**Pull request by sbender9**
This is only supported if security is turned on.

This supports two name spaces, one for global data and one for user specific data. All data is stored in the JSON format.

For example, a client might want to store boat specific gauge configuration globally so that other users have access to it.

Otherwise, it could use the user area to store user specific preferences.

Global storage is at: `/signalk/v1/applicationData/global/:appid/:version`
User storage is at: `/signalk/v1/applicationData/user/:appid/:version`

There are two ways to update or add stored data: 

* You can POST any json data to any path: 

```
POST /signalk/v1/applicationData/user/my-application/1.0/unitPreferences
{ 
  "shortDistance": "m",
  "longDistance": "km"
}
```

* You can also use json patch format (http://jsonpatch.com):

```
POST /signalk/v1/applicationData/user/my-application/1.0
[
  { "op": "add", "path": "/unitPreferences", "value": { "shortDistace": "m" } },
  { "op": "add", "path": "/unitPreferences/longDistance", "value": "km"}
]
```

Use an HTTP GET request to retrieve data from the server:

`GET /signalk/v1/applicationData/user/my-application/1.0/unitPreferences/shortDistance`

You can just GET the list of keys:
```
GET /signalk/v1/applicationData/user/my-application/1.0/unitPreferences?keys=true
[ "longDistance", "shortDistance"]
```

You get can a list of available versions:
```
GET /signalk/v1/applicationData/user/my-application
[ "1.0", "1.1"]
```

### CODE REVIEW

**Comment by sbender9** [comment]
Replaces #643

File: test/applicationData.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    var result = await fetch(
+      `${url}/signalk/v1/applicationData/user/${APP_ID}/:${APP_VERSION}`,
+      {
+        method: 'POST',
+        headers: {
+          ...readHeaders,
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify({
+          setting: 'on'
```
Comment (tkurki) [inline test/applicationData.js]: All other tests use the same payload, is there something special here?

File: test/applicationData.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        })
+      }
+    )
+    result.status.should.equal(401)
```
Comment (tkurki) [inline test/applicationData.js]: What if the code actually writes, just returns 401?

File: test/applicationData.js [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  })
+
+  it('post user data works', async function () {
+    var result = await fetch(
```
Comment (tkurki) [inline test/applicationData.js]: There's so much boilerplate in the tests that the beef gets lost in the fluff. 

What if we had `post(globalOrUser, token, payload)` and `get(globalOrUser, token)` - the test would be all about what we are trying and the asserts?

All tests are run with the same server, right? Are you sure that a later test is not reading something that an earlier test has written by accident? There could something changed in the payload of each test?

File: test/applicationData.js [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    })
+  })
+
+  after(async function () {
+    await server.stop()
+  })
+
+  it('fetch global returns empty data', async function () {
+    var result = await fetch(
+      `${url}/signalk/v1/applicationData/global/${APP_ID}/:${APP_VERSION}`,
```
Comment (tkurki) [inline test/applicationData.js]: An extra : before the version?  

Should we enforce some format for the version and app id strings? maxlength, alphanumeric for appid, semantic version format for version?

File: test/servertestutilities.js [hunk 5]
```diff
@@ -61,16 +63,113 @@ WsPromiser.prototype.send = function (message) {
   })
 }
 
+const defaultSecurityConfig = {
+  allow_readonly: false,
+  expiration: '1d',
+  secretKey:
+    '3ad6c2b567c43199e1afd2307ef506ea9fb5f8becada1f86c15213d75124fbaf4647c3f7202b788bba5c01c8bb8fdc52e8ca5bd484be36b6900ac03b88b6063b6157bee1e638acde1936d6ef4717884de63c86e9f50c8ee12b15bf837268b04bc09a461f5dddaf71dfc7205cc549b29810a31515b21d57ac5fdde29628ccff821cfc229004c4864576eb7c238b0cd3a6d774c14854affa1aeedbdb1f47194033f18e50d9dc1171a47e36f26c864080a627c500d1642fc94f71e93ff54022a8d4b00f19e88a0610ef70708ac6a386ba0df7cab201e24d3eb0061ddd0052d3d85cda50ac8d6cafc4ecc43d8db359a85af70d4c977a3d4b0d588f123406dbd57f01',
```
Comment (tkurki) [inline test/servertestutilities.js]: Should this be just a random string, like timestamp? Even if this is test code.

File: src/interfaces/applicationData.js [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  function saveApplicationData(req, appid, version, isUser, data, callback) {
+    const applicationDataDir = path.join(
+      app.config.configPath,
+      'applicationData'
+    )
+
+    if (!fs.existsSync(applicationDataDir)) {
+      fs.mkdirSync(applicationDataDir)
```
Comment (tkurki) [inline src/interfaces/applicationData.js]: handle error gracefully here? may not have permission

File: src/interfaces/applicationData.js [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    let location = path.join(
+      app.config.configPath,
+      'applicationData',
+      isUser ? req.skPrincipal.identifier : 'global'
+    )
+
+    return path.join(location, appid)
+  }
+
+  function pathForApplicationData(req, appid, version, isUser) {
```
Comment (tkurki) [inline src/interfaces/applicationData.js:201]: Should validate appid and versio here. As it is is directory traversal possible?

File: src/interfaces/applicationData.js [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    return path.join(
+      dirForApplicationData(req, appid, isUser),
+      `${version}.json`
+    )
+  }
+
+  function saveApplicationData(req, appid, version, isUser, data, callback) {
```
Comment (tkurki) [inline src/interfaces/applicationData.js:208]: validation also here

File: test/servertestutilities.js [hunk 9]
```diff
@@ -61,16 +63,113 @@ WsPromiser.prototype.send = function (message) {
...
+      props.config.settings.security = {
+        strategy: './tokensecurity'
+      }
+      props.securityConfig = {
+        ...defaultSecurityConfig,
+        ...(securityConfig || {})
+      }
+    }
+
+    process.env.SIGNALK_NODE_CONDFIG_DIR = require('path').join(
```
Comment (tkurki) [inline test/servertestutilities.js]: Use the variable name without the typo.

File: test/applicationData.js [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+      {
+        method: 'POST',
+        headers: {
+          ...readHeaders,
+          'Content-Type': 'application/json'
+        },
+        body: JSON.stringify(TEST_SETTINGS)
+      }
+    )
+    result.status.should.equal(401)
```
Comment (tkurki) [inline test/applicationData.js]: Should this be 403 Forbidden, not 401 Unauthorized?

**Comment by sbender9** [comment]
I can't figure out the test failure. It doesn't fail if you just run that one test...

File: test/applicationData.js (hunk 10 above)
Comment (sbender9) [reply test/applicationData.js]: Probably, but I don't think this PR is the best place to fix it

**Comment by sbender9** [comment]
@tkurki I think I've addressed all your comments

**Comment by sbender9** [comment]
Oh. Missed the comment about multiple apps and versions. I'll add those...

**Pull request merged**